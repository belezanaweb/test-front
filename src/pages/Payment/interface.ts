interface IProductDataPage {
  discount: number
  id: string
  shippingTotal: number
  subTotal: number
  total: number
}

export interface IPaymentPage {
  productData: IProductDataPage
  handleSubmit: any
}
