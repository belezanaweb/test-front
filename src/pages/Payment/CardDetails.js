import React, { useContext } from 'react'
import CardDetailsContext from '../../contexts/CardDetails'
import { CardDetailsWrapper, Content, ErrorMessage, Input, Label, LastDatails } from './style'

const CardDetails = () => {
    const { form, handleInputChange } = useContext(CardDetailsContext)

    return (
        <CardDetailsWrapper>
            <Content>
                <Label>Número do cartão</Label>
                <Input
                    placeholder="____.____.____.____"
                    onChange={handleInputChange}
                    name="cardNumber"
                    value={form.cardNumber}
                    mask="9999.9999.9999.9999"
                    pattern="^\d{4}\.\d{4}\.\d{4}\.\d{4}$"
                    required
                />
                {(!form.cardNumber ||
                    !/\d{4}\.\d{4}\.\d{4}\.\d{4}/.test(form.cardNumber)) &&
                    <ErrorMessage>Invalid Field</ErrorMessage>}
            </Content>
            <Content>
                <Label>Nome do Titular</Label>
                <Input
                    placeholder="Como no cartão"
                    onChange={handleInputChange}
                    name="cardholderName"
                    value={form.cardholderName}
                    pattern="[A-Za-zÀ-ú ]{5,}"
                    required
                />
                {(!form.cardholderName ||
                    !/[A-Za-zÀ-ú ]{5,}/.test(form.cardholderName)) &&
                    <ErrorMessage>Invalid Field</ErrorMessage>}
            </Content>
            <LastDatails>
                <Content>
                    <Label>Validade (mês/ano):</Label>
                    <Input
                        placeholder="__/____"
                        onChange={handleInputChange}
                        name="expiringDate"
                        value={form.expiringDate}
                        mask="99/9999"
                        pattern="^(1[0-2]|0[1-9]|\d)/(20[2-9][0-9]|[2-9][1-9])$"
                        required
                    />
                    {(!form.expiringDate ||
                        !/(1[0-2]|0[1-9]|\d)-(20[2-9][0-9]|[2-9][1-9])/.test(form.expiringDate)) &&
                        <ErrorMessage>Invalid Field</ErrorMessage>}
                </Content>
                <Content>
                    <Label>CVV:</Label>
                    <Input
                        placeholder="___"
                        onChange={handleInputChange}
                        name="cvv"
                        value={form.cvv}
                        mask="999"
                        pattern="[0-9]{3}"
                        required
                    />
                    {(!form.cvv ||
                        !/[0-9]{3}/.test(form.cvv)) &&
                        <ErrorMessage>Invalid Field</ErrorMessage>}
                </Content>
            </LastDatails>
        </CardDetailsWrapper>
    )
}

export default CardDetails