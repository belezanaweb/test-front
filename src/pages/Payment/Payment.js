import React, { useEffect, useState } from 'react'
import {
  Button,
  Container,
  Grid,
  makeStyles,
  Paper,
  TextField,
  Typography
} from '@material-ui/core'
import Header from '../../components/Header'
import SumOfValues from '../../components/SumOfValues'
import ValidateCreditCard from '../../services/ValidateCreditCard'
import * as moment from 'moment'
import InputMask from 'react-input-mask'

const useStyles = makeStyles({
  root: {
    alignItems: 'center',
    marginTop: '10px'
  },
  button: {
    backgroundColor: '#FF7800',
    color: '#fff',
    fontWeight: 'bold',
    width: '100%'
  },
  buttonContainer: {
    padding: '0px',
    marginTop: '10px'
  },
  paper: {
    borderRadius: '3px',
    color: '#fff',
    padding: '12px'
  },
  form: {
    '& label.Mui-focused': {
      color: '#A43287'
    },
    '& .MuiInput-underline:after': {
      borderBottomColor: '#A43287'
    },
    '& .MuiOutlinedInput-root': {
      '&.Mui-focused fieldset': {
        borderColor: '#A43287'
      }
    },
    '& .MuiFormControl-root': {
      marginBottom: '25px'
    }
  }
})

function Payment() {
  const classes = useStyles()
  const listProducts = JSON.parse(localStorage.getItem('listProducts'))
  const [disabledButton, setDisabledButton] = useState(true)
  const [creditCardNumber, setCreditCardNumber] = useState()
  const [creditCardName, setCreditCardName] = useState()
  const [creditCardValidate, setCreditCardValidate] = useState()
  const [creditCardCVV, setCreditCardCVV] = useState()
  const [errorCreditCardNumber, setErrorCreditCardNumber] = useState(false)
  const [errorCreditCardName, setErrorCreditCardName] = useState(false)
  const [errorCreditCardValidate, setErrorCreditCardValidate] = useState(false)
  const [errorCreditCardCVV, setErrorCreditCardCVV] = useState(false)

  const handleChangeCreditCard = (event) => {
    localStorage.setItem('creditCardNumber', event.target.value)
    setCreditCardNumber(event.target.value)
    event.preventDefault()
    const cardNumberValidate = ValidateCreditCard(event.target.value)
    if (!cardNumberValidate && String(creditCardNumber).length !== 0) {
      setErrorCreditCardNumber(true)
    } else {
      setErrorCreditCardNumber(false)
    }
  }

  const handleChangeCardName = (event) => {
    localStorage.setItem('creditCardName', event.target.value)
    setCreditCardName(event.target.value)
    event.preventDefault()
    if (String(event.target.value).length === 0 || String(event.target.value).trim() === '') {
      setErrorCreditCardName(true)
    } else {
      setErrorCreditCardName(false)
    }
  }

  const handleChangeCardValidate = (event) => {
    localStorage.setItem('creditCardValidate', event.target.value)
    setCreditCardValidate(event.target.value)
    event.preventDefault()
    if (!moment(String(event.target.value), 'MM/YYYY', true).isValid()) {
      setErrorCreditCardValidate(true)
    } else {
      setErrorCreditCardValidate(false)
    }
  }

  const handleChangeCardCVV = (event) => {
    localStorage.setItem('creditCardCVV', event.target.value)
    setCreditCardCVV(event.target.value)
    event.preventDefault()
    if (String(event.target.value).length < 3 || String(event.target.value).length > 4) {
      setErrorCreditCardCVV(true)
    } else {
      setErrorCreditCardCVV(false)
    }
  }

  useEffect(() => {
    if (
      creditCardNumber !== undefined &&
      creditCardName !== undefined &&
      creditCardValidate !== undefined &&
      creditCardCVV !== undefined
    ) {
      if (
        !errorCreditCardNumber &&
        !errorCreditCardName &&
        !errorCreditCardValidate &&
        !errorCreditCardCVV
      ) {
        setDisabledButton(false)
      }
    }
  }, [
    creditCardNumber,
    creditCardName,
    creditCardValidate,
    creditCardCVV,
    errorCreditCardNumber,
    errorCreditCardName,
    errorCreditCardValidate,
    errorCreditCardCVV
  ])
  return (
    <div>
      <Header select={2} />
      <Container className={classes.root}>
        <Typography style={{ color: '#999', fontWeight: 'bold' }}>CARTÃO DE CRÉDITO</Typography>
        <Paper elevation={3} className={classes.paper}>
          <form className={classes.form}>
            <InputMask
              mask="9999.9999.9999.9999"
              value={creditCardNumber}
              disabled={false}
              maskChar=" "
              onChange={handleChangeCreditCard}
            >
              {() => (
                <TextField
                  id="creditCardNumber"
                  label="Número do cartão:"
                  required
                  fullWidth
                  variant="outlined"
                  error={errorCreditCardNumber}
                  helperText={errorCreditCardNumber && 'Número incompleto ou inválido.'}
                />
              )}
            </InputMask>
            <TextField
              required
              id="creditCardName"
              label="Nome do Titular:"
              variant="outlined"
              placeholder="Como no cartão"
              fullWidth
              value={creditCardName}
              onChange={handleChangeCardName}
              error={errorCreditCardName}
            />
            <Grid container spacing={3}>
              <Grid item xs={7}>
                <InputMask
                  mask="99/9999"
                  value={creditCardValidate}
                  disabled={false}
                  maskChar=" "
                  onChange={handleChangeCardValidate}
                >
                  {() => (
                    <TextField
                      id="creditCardValidate"
                      label="Validade (mês/ano):"
                      required
                      fullWidth
                      variant="outlined"
                      error={errorCreditCardValidate}
                      helperText={errorCreditCardValidate && 'Data inválida.'}
                    />
                  )}
                </InputMask>
              </Grid>
              <Grid item xs={5}>
                <TextField
                  required
                  id="creditCardCVV"
                  label="CVV:"
                  variant="outlined"
                  placeholder="___"
                  fullWidth
                  value={creditCardCVV}
                  onChange={handleChangeCardCVV}
                  error={errorCreditCardCVV}
                  helperText={errorCreditCardCVV && 'Número incompleto ou inválido.'}
                />
              </Grid>
            </Grid>
          </form>
        </Paper>
        <SumOfValues productList={listProducts} />
        <div className={classes.buttonContainer}>
          <Button
            id="buttonPayment"
            type="submit"
            className={classes.button}
            variant="contained"
            size="large"
            href="/sucesso"
            disabled={disabledButton}
          >
            FINALIZAR O PEDIDO
          </Button>
        </div>
      </Container>
    </div>
  )
}
export default Payment
