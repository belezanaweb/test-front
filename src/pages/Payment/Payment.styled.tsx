import styled from 'styled-components/macro'

export const ProductsTotalWrapper = styled.div`
  @media (min-width: 768px) {
    width: 40%;

    section {
      margin-top: 0;
    }
  }
`
