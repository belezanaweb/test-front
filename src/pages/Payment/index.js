import React from 'react'

const Payment = () => {
  return (
    <div>
      <h1>Payment</h1>
    </div>
  )
}

export default Payment
