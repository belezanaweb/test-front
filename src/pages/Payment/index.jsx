import React, { useState } from 'react'
import { Container } from './styles'
import { useDispatch } from 'react-redux'
import { handlePayment } from '../../redux/slices/paymentSlice'

import loadable from '@loadable/component'

const Header = loadable(() => import('../../components/Header'), {
  resolveComponent: (components) => components.Header
})
const PaymentForm = loadable(() => import('../../components/PaymentForm'), {
  resolveComponent: (components) => components.PaymentForm
})
const Summary = loadable(() => import('../../components/Summary'), {
  resolveComponent: (components) => components.Summary
})
const ActionButton = loadable(() => import('../../components/ActionButton'), {
  resolveComponent: (components) => components.ActionButton
})

export const Payment = () => {
  const [cardNumber, setCardNumber] = useState('')
  const [name, setName] = useState('')
  const [expiryDate, setExpiryDate] = useState('')
  const [cvv, setCvv] = useState('')
  const dispatch = useDispatch()

  const handlePaymentInformation = () => {
    const payload = {
      cardNumber,
      name,
      expiryDate,
      cvv
    }
    dispatch(handlePayment(payload))
  }

  return (
    <>
      <Header />
      <Container>
        <PaymentForm
          fillCardNumber={setCardNumber}
          fillName={setName}
          fillCVV={setCvv}
          fillExpiryDate={setExpiryDate}
        />
        <Summary />
        <ActionButton
          handlePaymentInformation={handlePaymentInformation}
          pathToGo={'/confirmation'}
          actionText={'Finalizar o pedido'}
        />
      </Container>
    </>
  )
}
