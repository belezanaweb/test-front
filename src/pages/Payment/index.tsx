import React from 'react'
import { useNavigate } from 'react-router-dom'

import { useCreditCard } from '../../hooks/creditCard'
import { useCart } from '../../hooks/cart'

import NavBar from '../../components/NavBar'
import CreditCardForm from '../../components/CreditCardForm'
import Total from '../../components/Total'
import { Button } from '../../components/Button'

import { Root, Container, Content } from './styles'

const Payment: React.FC = () => {
  const navigate = useNavigate()
  const { totalData } = useCart()
  const { formIsValid } = useCreditCard()

  const handleSubmit = () => {
    if(formIsValid) navigate('/confirmacao')
  }

  return (
    <Root>
      <NavBar actualStep={'payment'} />
      <Container>
        <Content>
          <h1>cartão de crédito</h1>
          <CreditCardForm
            formId={'creditCardForm'}
            formSubmitted={handleSubmit}
          />
          <Total
            subTotal={totalData.subTotal}
            shipping={totalData.shippingTotal}
            discount={totalData.discount}
            total={totalData.total}
          />
          <Button type={'submit'} form={'creditCardForm'} disabled={!formIsValid} >
            finalizar o pedido
          </Button>
        </Content>
      </Container>
    </Root>
  )
}

export default Payment
