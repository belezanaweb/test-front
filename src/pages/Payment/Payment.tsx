import React from 'react'

import { IPaymentPage } from './interface'

import Button from '../../components/Button'
import CartTable from '../../components/CartTable'

import { FormBuilder } from '../../components/Form'
import { fields } from '../../components/Form/data'
import Title from '../../components/Title'
import * as S from './Payment.styled'

export const Payment = ({ productData, handleSubmit }: IPaymentPage) => {
  return (
    <>
      <Title>Cartão de crédito</Title>
      <FormBuilder onSubmit={handleSubmit} fieldsList={fields}>
        <S.ProductsTotalWrapper>
          <CartTable
            subTotal={productData.subTotal}
            shippingTotal={productData.shippingTotal}
            discount={productData.discount}
            total={productData.total}
          />
          <Button type="submit">Finalizar o pedido</Button>
        </S.ProductsTotalWrapper>
      </FormBuilder>
    </>
  )
}
