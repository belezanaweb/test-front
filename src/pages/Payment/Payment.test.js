import React from 'react'
import { render } from '@testing-library/react'

import Payment from './Payment'

describe('Payment Page', () => {
  test('Renders with title', async () => {
    const { asFragment, getByText } = render(<Payment />)
    expect(getByText('Cartão de crédito')).toBeInTheDocument()

    expect(asFragment()).toMatchSnapshot()
  })
})
