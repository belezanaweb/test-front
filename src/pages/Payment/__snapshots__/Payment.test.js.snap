// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Payment Page Renders with title 1`] = `
<DocumentFragment>
  <h2
    class="Payment__StyledPageTitle-k343ez-0 hHjtHH"
  >
    Cartão de crédito
  </h2>
  <form>
    <div
      class="Payment__StyledBox-k343ez-1 dObvgT"
    >
      <div>
        <label
          class="Payment__StyledLabel-k343ez-2 fglfJn"
          for="cardNumber"
        >
          Número do cartão:
        </label>
        <input
          autocomplete="off"
          class="Payment__StyledInput-k343ez-3 ighMjI"
          id="cardNumber"
          maxlength="16"
          name="cardNumber"
          placeholder="____.____.____.____"
          value=""
        />
      </div>
      <div>
        <label
          class="Payment__StyledLabel-k343ez-2 fglfJn"
          for="name"
        >
          Nome do Titular:
        </label>
        <input
          class="Payment__StyledInput-k343ez-3 ighMjI"
          id="name"
          name="name"
          placeholder="Como no cartão"
          value=""
        />
      </div>
      <div>
        <label
          class="Payment__StyledLabel-k343ez-2 fglfJn"
          for="date"
        >
          Validade mês/ano):
        </label>
        <input
          class="Payment__StyledInput-k343ez-3 ighMjI"
          id="date"
          maxlength="7"
          name="date"
          placeholder="_ _ / _ _ _ _"
          value=""
        />
      </div>
      <div>
        <label
          class="Payment__StyledLabel-k343ez-2 fglfJn"
          for="cvv"
        >
          CVV:
        </label>
        <input
          autocomplete="off"
          class="Payment__StyledInput-k343ez-3 ighMjI"
          id="cvv"
          maxlength="3"
          name="cvv"
          placeholder="_ _ _"
          value=""
        />
      </div>
    </div>
    <div
      class="Total__StyledTotal-sc-4ywkhv-0 hmXNNq"
    >
      <p
        class="Total__StyledTotalRow-sc-4ywkhv-1 eHCEJD"
      >
        <span>
          PRODUTOS
        </span>
        <span>
          R$ 624,80
        </span>
      </p>
      <p
        class="Total__StyledTotalRow-sc-4ywkhv-1 eHCEJD"
      >
        <span>
          FRETE
        </span>
        <span>
          R$ 5,30
        </span>
      </p>
      <p
        class="Total__StyledTotalRow-sc-4ywkhv-1 Total__StyledTotalRowDiscount-sc-4ywkhv-2 eHCEJD jxNTuf"
      >
        <span>
          DESCONTO
        </span>
        <span>
          - R$ 30,00
        </span>
      </p>
      <p
        class="Total__StyledTotalRow-sc-4ywkhv-1 Total__StyledTotalRowTotal-sc-4ywkhv-3 eHCEJD biziwy"
      >
        <span>
          TOTAL
        </span>
        <span>
          R$ 600,10
        </span>
      </p>
    </div>
    <button
      class="Button__StyledButton-sc-1vikqwd-0 dNzGGX"
      disabled=""
      type="submit"
    >
      Finalizar o pedido
    </button>
  </form>
</DocumentFragment>
`;
